src/crowdin_client.cpp

```diff
@@ -135,150 +159,332 @@ CrowdinClient::~CrowdinClient() {}
+                                                   const std::string& lang_tag,
                                                    const std::wstring& output_file)
 {
-    // NB: "export_translated_only" means the translation is not filled with the source string
-    //     if there's no translation, i.e. what Poedit wants.
-    auto url = "/api/project/" + project_id + "/export-file"
-                   "?json="
-                   "&export_translated_only=1"
-                   "&language=" + lang.LanguageTag() +
-                   "&file=" + http_client::url_encode(file);
-    return m_api->download(url, output_file);
+    cout << "\n\nGetting file URL: " << "/projects/" + std::to_string(project_id) + "/files/" + std::to_string(file_id) + "/download" << "\n\n";
+    return RefreshToken().then([=] () {
+    return m_api->post(
+        "projects/" + std::to_string(project_id) + "/translations/builds/files/" + std::to_string(file_id),
+        json_data(json({
+            { "targetLanguageId", lang_tag },
+            // for XLIFF files should be exported "as is" so set to `false`
+            { "exportAsXliff", !file_name.Lower().EndsWith(".xliff.xliff") }
+        })))
+        .then([this, output_file] (json r) {
+            string url = r["data"]["url"],
+                   proto(wxString(url).BeforeFirst(':').mb_str()),
+                   host(wxString(url).AfterFirst('/').AfterFirst('/').BeforeFirst('/').mb_str());
+
+            auto downloader = make_shared<crowdin_http_client>(*this, proto + "://" + host);
+            cout << "\n\nGotten file URL: "<< r << "\n\n";
+            return downloader->download(url, output_file)
+                .then([downloader]() {
+                    return dispatch::make_ready_future();
+                });
+        });
+    });
 }
 
 
-dispatch::future<void> CrowdinClient::UploadFile(const std::string& project_id,
-                                                 const std::wstring& file,
-                                                 const Language& lang,
+dispatch::future<void> CrowdinClient::UploadFile(const long project_id,
+                                                 const long file_id,
+                                                 const std::string& lang_tag,
                                                  const std::string& file_content)
 {
-    auto url = "/api/project/" + project_id + "/upload-translation";
-
-    multipart_form_data data;
-    data.add_value("json", "");
-    data.add_value("language", lang.LanguageTag());
-    data.add_value("import_duplicates", "0");
-    data.add_value("import_eq_suggestions", "0");
-    data.add_file("files[" + str::to_utf8(file) + "]", "upload.po", file_content);
-
-    return m_api->post(url, data).then([](json){});
+    return RefreshToken().then([=] () {
+    return m_api->post(
+            "storages",
+            octet_stream_data(file_content),
+            { { "Crowdin-API-FileName", L"poedit.xliff"} }
```

Comment:
No problem with that as just double checked with Crowdin API v2 developers. File is never overwritten since completely separate independent storage created per each such request. While filename is used only for informative/debugging purpose plus to detect it type (e.g. when to put e.g. "poedit.txt" - it will be rejected by e.g. [Upload Translations](https://support.crowdin.com/api/v2/#operation/api.projects.translations.postOnLanguage) request) if it source file is not `.txt` as well.